### ISSUE TITLE
Non-WSI extensions are are broken

### ISSUE THREAD

**Issue by corngood** [expipiplus1/vulkan#19]
e.g. EXT_debug_report

Only the core APIs and WSI extensions actually have symbols provided in the loader.  Everything else needs to be dynamically loaded via Get[...]ProcAddr.

I threw something together as a proof of concept, which gets the extension working for me: @93cea16 (WIP).

What do you think of this approach?  It might even make sense to fully transistion to dynamic loading, using e.g. `dlopen` to get vkGetInstanceProcAddr, and grabbing everything else from there.  However in that case you'd need to handle failure, which would probably rule out the unsafePerformIO.

With this change users would be encouraged reuse command functions after partially applying them on instance / device, which is equivalent to good practice in C.

**Comment by expipiplus1** [comment]
Ah, good catch, thanks.

What would the advantage be for grabbing everything with `dlopen`?

This looks like a good start, however it could behave a little better if `getInstanceProcAddr` returns a null pointer (I'm sure that's just an artifact of this being a WIP :wink:). I think that the most 'Haskelly' value for runtime loaded functions such as `vkDebugReportMessageEXT` is `VkInstance -> Maybe (A -> B -> IO C)` or `VkInstance -> IO (Maybe (A -> B -> IO C))`. Although it's a little more inconvenient to use it best represents the API supplied by Vulkan. 

I imagine in a high level Vulkan wrapper the user could specify up front which extensions they want to use and the wrapper tries to load the requisite functions straight after instance/device initialization, putting the functions into the instance/device record on success and failing to create the instance/device on failure.

**Comment by corngood** [comment]
From the [loader docs](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/loader/LoaderAndLayerInterface.md):

> The loader library on Windows, Linux and Android will export all core Vulkan and all appropriate Window System Interface (WSI) extensions. This is done to make it simpler to get started with Vulkan development.

> Applications are not required to link directly to the loader library, instead they can use the appropriate platform specific dynamic symbol lookup on the loader library to initialize the application’s own dispatch table. This allows an application to fail gracefully if the loader cannot be found, and it provides the fastest mechanism for the application to call Vulkan functions. An application will only need to query (via system calls such as dlsym()) the address of vkGetInstanceProcAddr from the loader library. Using vkGetInstanceProcAddr the application can then discover the address of all instance and global functions and extensions, such as vkCreateInstance, vkEnumerateInstanceExtensionProperties and vkEnumerateInstanceLayerProperties in a platform independent way.

> VkGetDeviceProcAddr is particularly interesting because it will provide the most efficient way to call into the ICD.

So theoretically you're going to get better performance from the more direct entry points.  Haskell just happens to have a fairly elegant way of doing this with:

```
let createRenderPass = vkCreateRenderPass device
```

I actually don't think the low level bindings needs to handle missing functions (i.e. with Maybe).  You're already opting in to each extension.  Higher level bindings can do some type magic to make it safe.

Even if you move to loading it dynamically, I think checking the availability of each function with Maybe would be too cumbersome.

**Comment by expipiplus1** [comment]
I had a thought on this: I think that some kind of "device" helper would be useful, it contains all the device functions and can be created once.

```haskell
data VkDeviceFun = VkDeviceFun
  { vkCreateRenderPass :: ...
  , vkDestroyRenderPass :: ...
  }

hasFunction :: VkDeviceFun -> String -> Bool
```

The same goes for instances.

Your solution with `unsafePerformIO` is very nice too.

**Comment by expipiplus1** [comment]
I have implemented a basic version of this in #31
